### DISCUSSION TITLE
Tiles layer performance issue with layer update.

### DISCUSSION THREAD

**Discussion by peaceshi** [visgl/deck.gl#5891] [General]
#### Description
Tiles layer performance issue with layer update.

#### Repro Steps
Github: [source code]（https://github.com/peaceshi/tiles-layer-test/deployments/activity_log?environment=Production） that reproduces the behavior
Vercel: [example](https://tiles-layer-test.vercel.app/)

#### Environment
 - Framework Version: deck.gl 8.4.17
 - Browser Version: Microsoft Edge  Chromium 
 - OS: All

Seems like should not use `!==` to compare to `Objects`, because of that will always return ture. 
Only primitive type can use it.
![image](https://user-images.githubusercontent.com/10380828/122260339-bbf34300-cf05-11eb-927e-64d4c633c41d.png)

https://github.com/visgl/deck.gl/blob/38149ae04f9af93d5dececf11ce2cdc551e6ba8f/modules/react/src/deckgl.js#L79

Then it will be `forceUpdate` the whole react dom every time when drag or move because it makes a new viewport.
And redraw redraw and redraw frequently that causes a bad performance.
https://github.com/visgl/deck.gl/blob/38149ae04f9af93d5dececf11ce2cdc551e6ba8f/modules/react/src/deckgl.js#L83

So that ,when call rebuildViewports(), we set `this._viewports.length = 0` maybe a way to fix it.
https://github.com/visgl/deck.gl/blob/38149ae04f9af93d5dececf11ce2cdc551e6ba8f/modules/core/src/lib/view-manager.js#L295

**Comment by peaceshi** [comment]
https://github.com/visgl/deck.gl/pull/5884

**Comment by peaceshi** [comment]
https://github.com/peaceshi/deck.gl/issues/1

**Comment by peaceshi** [comment]
if we use a big cache number, `cache.values()` will be very huge.

does it have a better way ?

https://github.com/visgl/deck.gl/blob/732a1c0ac6f0c3a47634023d69dca34b17b7d442/modules/geo-layers/src/tile-layer/tileset-2d.js#L315

**Comment by Pessimistress** [comment]
Rerender is always required when viewports change, however frequently it is. If you don't rerender then you won't see any visual feedback during drag and zoom.

**Comment by Pessimistress** [comment]
Before jumping to conclusions, I suggest that you describe how you are using the TileLayer, and if possible, the relevant code. It’s very possible that you can simply tweak your own usage pattern to improve performance. If you do think a particular operation in the render cycle is causing the problem, you can use the profiler tool to demonstrate that it's taking significant CPU time.

**Comment by peaceshi** [comment]
> Before jumping to conclusions, I suggest that you describe how you are using the TileLayer, and if possible, the relevant code. It’s very possible that you can simply tweak your own usage pattern to improve performance. If you do think a particular operation in the render cycle is causing the problem, you can use the profiler tool to demonstrate that it's taking significant CPU time.

Sorry for that it was not caused by the viewport array.

Our team member makes this image. 
Source code here: https://github.com/peaceshi/tiles-layer-test/blob/2390ba6146e74f65356dc3cb0c7e2cc85eedb73e/src/App.js#L54

![2](https://user-images.githubusercontent.com/10380828/122408169-c0c30000-cfb4-11eb-80a5-0014d1b24d3a.jpg)

After this profiling, `fetchTileData` has been changed to use async/await. 
`deck.gl` updated to `8.5.0-alpha.10`.
Sample site also updated : https://tiles-layer-test.vercel.app/
![image](https://user-images.githubusercontent.com/10380828/122412998-9d01b900-cfb8-11eb-910f-a5ad768883cc.png)

You can see the fps drop very clearly when drag/move.

**Reply by Pessimistress** [reply]
I don't see any issue in your performance profile. If you pan to an area that has not been loaded, then the layer is expected to create new sub layers as new tiles come into view. The only optimization I would suggest is to use the `extent` prop to prevent empty tiles from being created, instead of rendering empty bitmap layers.

I don't see any noticeable lag on your deployed app either. Are you sure it's not a device-specific issue?

**Reply by jcready** [reply]
One thing that we noticed about the TileLayer is that it attempts to fetch new tiles immediately when the view changes which was causing very slight blocking on the main thread. It wasn't as important for our use case to load tiles as fast as possible so we extended the TileLayer such that the requests for new tiles was essentially debounced. Perhaps there is a better way to do this, but this seemed to work for us:

```ts
export DebouncedTileLayer extends TileLayer {
  _updateTileset() {
    if (!this.state.updateTimeout) {
      this.state.updateTimeout = 1;
      super._updateTileset();
    } else {
      clearTimeout(this.state.updateTimeout);
      this.state.updateTimeout = setTimeout(() => super._updateTileset(), 500);
    }
  }
}
```

The difference is especially noticeable when using something like `FlyToInterpolator` for viewport transitions or if you have several TileLayer instances.

**Reply by jfuehner** [reply]
@jcready Thanks for sharing your DebouncedTileLayer. This small change improved our tile layer loading experience as well.
